libcloud/compute/drivers/ec2.py

```diff
@@ -4402,6 +4498,127 @@ def delete_image(self, image):
+
+            if limit and len(ri_offerings) >= limit:
+                ri_offerings = ri_offerings[:limit]
+                break
+
+        return ri_offerings
+
+    def _ex_iterate_reserved_instances_offerings(self, params):
+        next_page_token = ''
+
+        while next_page_token is not None:
+            response = self.connection.request(self.path, params=params).object
+            next_page_token = findtext(response, xpath='nextToken', namespace=NAMESPACE) or None
+            yield self._to_reserved_instances_offerings(response)
+            params['NextToken'] = next_page_token
+
+    def _to_reserved_instances_offerings(self, response):
+        return [self._to_reserved_instance_offering(el) for el in response.findall(
+            fixxpath(xpath='reservedInstancesOfferingsSet/item', namespace=NAMESPACE))
+        ]
+
+    def _to_reserved_instance_offering(self, element):
+        params = self._get_extra_dict(element, RESERVED_INSTANCES_OFFERINGS_MAP)
+
+        recurring_charges = []
+        for el in element.findall(fixxpath(xpath='recurringCharges/item', namespace=NAMESPACE)):
+            recurring_charges.append({'frequency': findtext(element=el,
+                                                            xpath='frequency',
+                                                            namespace=NAMESPACE),
+                                      'amount': findtext(element=el,
+                                                         xpath='amount',
+                                                         namespace=NAMESPACE)
+                                      })
+        params['recurring_charges'] = recurring_charges
+        params['region'] = self.region_name
+
+        return EC2ReservedInstancesOffering(**params)
+
+    def ex_purchase_reserved_instances_offering(self, offering_id, instance_count,
+                                                limit_price=None, dry_run=False):
+        """
+        Purchases a Reserved Instance for use with your account.
+
+        :param offering_id: The ID of the Reserved Instance offering to purchase.
+        :type  offering_id: str
+
+        :param instance_count: The number of Reserved Instances to purchase.
+        :type instance_count: int
+
+        :param limit_price: Specified for Reserved Instance Marketplace offerings to limit the
+            total order and ensure that the Reserved Instances are not purchased at unexpected
+            prices.
+        :type limit_price: float
+
+        :param dry_run: Checks whether you have the required permissions for the action,
+            without actually making the request, and provides an error response.
+            If you have the required permissions, the error response is DryRunOperation.
+            Otherwise, it is UnauthorizedOperation.
+        :type dry_run: bool
+
+        :return: dict with 'request_id' and 'reserved_instance_id' in case of successful operation.
```

Comment:
@uty томущо ` in case of unsuccessful` буде Exception.

Reply:
@vaniakov, тогда это отсюда нужно убрать, т.к. `return` всегда одинаковый.

Reply:
@uty що забрати?

Reply:
`in case of successful operation`. Ты  описываешь `return`, поэтому это выглядит так, будто в unsuccessful случае будет возвращаться что-то другое. Но на деле выбрасывается исключение и return нет.

Можно про исключение написать в докстринге выше.